### PR TITLE
New Instance Segmentation Apphub - SOLOv2

### PR DESCRIPTION
**Pull request by vbvg2008**
Both TF and Torch implementation can reach ~0.31 mAP

Notebook spell checked

Test successful

### CODE REVIEW

File: apphub/instance_segmentation/solov2/solov2_tf.py [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+                  eval_steps_per_epoch=None):
+    assert im_size % 32 == 0, "im_size must be a multiple of 32"
+    num_device = get_num_devices()
+    train_ds, val_ds = mscoco.load_data(root_dir=data_dir, load_masks=True)
+    batch_size = num_device * batch_size_per_gpu
+    pipeline = fe.Pipeline(
+        train_data=train_ds,
+        eval_data=val_ds,
+        test_data=val_ds,
+        batch_size=num_device * batch_size_per_gpu,
```
Comment (TortoiseHam) [inline apphub/instance_segmentation/solov2/solov2_tf.py:502]: why not use the batch_size from line 497 here?

File: apphub/instance_segmentation/solov2/solov2_tf.py [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+            Delete(keys="bbox"),
+            Delete(keys="image_id", mode="!test")
+        ],
+        pad_value=0,
+        num_process=8 * num_device)
+    init_lr = 1e-2 / 16 * batch_size
+    model = fe.build(model_fn=lambda: solov2(input_shape=(im_size, im_size, 3)),
+                     optimizer_fn=lambda: tf.optimizers.SGD(learning_rate=init_lr, momentum=0.9))
+    network = fe.Network(ops=[
+        Normalize(inputs="image", outputs="image", mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)),
```
Comment (TortoiseHam) [inline apphub/instance_segmentation/solov2/solov2_tf.py:529]: why not the Numpy normalize?

File: apphub/instance_segmentation/solov2/solov2_torch.py [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+                 num_grids=(40, 36, 24, 16, 12),
+                 scale_ranges=((1, 96), (48, 192), (96, 384), (192, 768), (384, 2048)),
+                 coord_sigma=0.05,
+                 sampling_ratio=4.0):
+        super().__init__(inputs=inputs, outputs=outputs, mode=mode)
+        self.im_size = im_size
+        self.num_grids = num_grids
+        self.scale_ranges = scale_ranges
+        self.coord_sigma = coord_sigma
+        self.sampling_ratio = sampling_ratio
```
Comment (TortoiseHam) [inline apphub/instance_segmentation/solov2/solov2_torch.py:239]: is self.sampling_ratio actually used anywhere?

File: apphub/instance_segmentation/solov2/solov2_torch.py [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+        return np.array(gt_match, dtype="int32")
+
+    @staticmethod
+    def remove_empty_gt(masks, bboxes):
+        num_objects = masks.shape[0]
+        non_empty_mask = np.sum(masks.reshape(num_objects, -1), axis=1) > 0
+        return masks[non_empty_mask], bboxes[non_empty_mask]
+
+
+class NormalizePermute(TensorOp):
```
Comment (TortoiseHam) [inline apphub/instance_segmentation/solov2/solov2_torch.py:293]: why is this not done as numpy?

File: apphub/instance_segmentation/solov2/solov2_torch.py [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+                  eval_steps_per_epoch=None):
+    assert im_size % 32 == 0, "im_size must be a multiple of 32"
+    num_device = get_num_devices()
+    train_ds, val_ds = mscoco.load_data(root_dir=data_dir, load_masks=True)
+    batch_size = num_device * batch_size_per_gpu
+    pipeline = fe.Pipeline(
+        train_data=train_ds,
+        eval_data=val_ds,
+        test_data=val_ds,
+        batch_size=num_device * batch_size_per_gpu,
```
Comment (TortoiseHam) [inline apphub/instance_segmentation/solov2/solov2_torch.py:575]: batch_size

File: apphub/instance_segmentation/solov2/solov2_tf.py (hunk 2 above)
Comment (vbvg2008) [reply apphub/instance_segmentation/solov2/solov2_tf.py:529]: it's faster to do it on gpu, because the cpu- gpu transfer type is uint8 instead of float32

File: apphub/instance_segmentation/solov2/solov2_torch.py (hunk 4 above)
Comment (vbvg2008) [reply apphub/instance_segmentation/solov2/solov2_torch.py:293]: same reason listed earlier

**Review by TortoiseHam — APPROVED** [review]
LG for now
